File: src/commands/private/Panels.ts
import {
    ActionRowBuilder,
    ApplicationCommandOptionType,
    ButtonBuilder,
    ButtonStyle,
    ContainerBuilder,
    MediaGalleryBuilder,
    MediaGalleryItemBuilder,
    MessageFlags,
    SectionBuilder,
    SeparatorBuilder,
    SeparatorSpacingSize,
    TextDisplayBuilder,
} from "discord.js";
import { config } from "@/config";
import { Command, CommandCategory, type CommandData, type CommandHandlerContext } from "@/structures/index";

export default class extends Command {
    public static override readonly data: CommandData = {
        name: "panels",
        category: CommandCategory.Private,
        description: "Send custom panels",
        guilds: [config.supportGuildId],
        ownerOnly: true,
        options: [
            {
                name: "name",
                type: ApplicationCommandOptionType.String,
                description: "The panel name",
                required: true,
            },
        ],
    };

    public override async handleCommand({ interaction }: CommandHandlerContext) {
        if (!interaction.inCachedGuild()) return;
        if (!interaction.channel?.isSendable()) return;

        await interaction.deferReply({ flags: MessageFlags.Ephemeral });

        let components;
        switch (interaction.options.getString("name", true)) {
            case "links":
                components = this.getLinksComponents();
                break;
            case "rules":
                components = this.getRulesComponents();
                break;
            default:
                await interaction.editReply("Invalid name!");
                return;
        }

        await interaction.channel.send({
            components,
            flags: MessageFlags.IsComponentsV2,
            allowedMentions: {
                parse: [],
            },
        });

        await interaction.editReply("Sent!");
    }

    private getLinksComponents() {
        const container = new ContainerBuilder()
            .setAccentColor(this.colors.primary.int)
            .addTextDisplayComponents(new TextDisplayBuilder().setContent(`## ${this.emojis.logo} Orion - Links`))
            .addMediaGalleryComponents(
                new MediaGalleryBuilder().addItems(new MediaGalleryItemBuilder().setURL(this.config.bannerURL)),
            )
            .addSectionComponents(
                new SectionBuilder()
                    .addTextDisplayComponents(
                        new TextDisplayBuilder().setContent(
                            "> Explore our hosting services or login in the dashboard.",
                        ),
                    )
                    .setButtonAccessory(
                        new ButtonBuilder()
                            .setStyle(ButtonStyle.Link)
                            .setURL(this.config.domainURL)
                            .setEmoji({ id: this.emojis.logoId })
                            .setLabel("Website / Dashboard"),
                    ),
            )
            .addSeparatorComponents(new SeparatorBuilder())
            .addSectionComponents(
                new SectionBuilder()
                    .addTextDisplayComponents(
                        new TextDisplayBuilder().setContent("> Use the panel to manage your servers."),
                    )
                    .setButtonAccessory(
                        new ButtonBuilder()
                            .setStyle(ButtonStyle.Link)
                            .setURL(this.config.panelURL)
                            .setEmoji({ name: "🔗" })
                            .setLabel("Panel"),
                    ),
            )
            .addSeparatorComponents(new SeparatorBuilder())
            .addSectionComponents(
                new SectionBuilder()
                    .addTextDisplayComponents(
                        new TextDisplayBuilder().setContent(
                            "> Read the documentation to learn how to use our hosting.",
                        ),
                    )
                    .setButtonAccessory(
                        new ButtonBuilder()
                            .setStyle(ButtonStyle.Link)
                            .setURL(this.config.docsURL)
                            .setEmoji({ name: "📚" })
                            .setLabel("Documentation"),
                    ),
            )
            .addSeparatorComponents(new SeparatorBuilder())
            .addSectionComponents(
                new SectionBuilder()
                    .addTextDisplayComponents(
                        new TextDisplayBuilder().setContent("> Get the status of our services in real-time."),
                    )
                    .setButtonAccessory(
                        new ButtonBuilder()
                            .setStyle(ButtonStyle.Link)
                            .setURL(this.config.statusURL)
                            .setEmoji({ id: this.emojis.propertyId })
                            .setLabel("Status"),
                    ),
            )
            .addSeparatorComponents(new SeparatorBuilder())
            .addSectionComponents(
                new SectionBuilder()
                    .addTextDisplayComponents(new TextDisplayBuilder().setContent("> Need help? Open a ticket!"))
                    .setButtonAccessory(
                        new ButtonBuilder()
                            .setStyle(ButtonStyle.Link)
                            .setURL(this.config.ticketsPanelURL)
                            .setEmoji({ id: this.emojis.discordId })
                            .setLabel("Open a ticket"),
                    ),
            );

        const legal = new ContainerBuilder()
            .setAccentColor(this.colors.primary.int)
            .addTextDisplayComponents(new TextDisplayBuilder().setContent(`## ${this.emojis.logo} Legal Links`))
            .addTextDisplayComponents(
                new TextDisplayBuilder().setContent(
                    `> Terms of Service: ${this.config.domainURL}/terms` +
                        `\n> Privacy Policy: ${this.config.domainURL}/privacy`,
                ),
            );

        return [container, legal];
    }

    private getRulesComponents() {
        const container = new ContainerBuilder()
            .setAccentColor(this.colors.primary.int)
            .addTextDisplayComponents(new TextDisplayBuilder().setContent(`## ${this.emojis.logo} Orion - Rules`))
            .addMediaGalleryComponents(
                new MediaGalleryBuilder().addItems(new MediaGalleryItemBuilder().setURL(this.config.bannerURL)),
            )
            .addTextDisplayComponents(
                new TextDisplayBuilder().setContent(
                    "## Civilized behavior" +
                        "\nTreat other members with respect, even if you disagree with them." +
                        "\nHate speech, harassment and discriminatory language are strictly forbidden." +
                        "\nPlease respect the decisions of moderators and administrators.",
                ),
            )
            .addSeparatorComponents(new SeparatorBuilder().setSpacing(SeparatorSpacingSize.Large))
            .addTextDisplayComponents(
                new TextDisplayBuilder().setContent(
                    "## Privacy" +
                        "\nPrivate messages (PMs) may not be used for advertising or spamming." +
                        "\nInvitations to other servers or the addition of bots are strictly forbidden." +
                        "\nMembers' personal information must not be shared without their consent.",
                ),
            )
            .addSeparatorComponents(new SeparatorBuilder().setSpacing(SeparatorSpacingSize.Large))
            .addTextDisplayComponents(
                new TextDisplayBuilder().setContent(
                    "## Security" +
                        "\nPasswords and other sensitive information must not be shared." +
                        "\nReport any suspicious behavior or fraudulent activity to the moderators.",
                ),
            )
            .addSeparatorComponents(new SeparatorBuilder().setSpacing(SeparatorSpacingSize.Large))
            .addTextDisplayComponents(
                new TextDisplayBuilder().setContent(
                    "## Other" +
                        "\nLinks to malicious websites or explicit content are prohibited." +
                        "\nExcessive use of capital letters, spam or animated emotes is prohibited." +
                        "\nMembers may not impersonate others.",
                ),
            )
            .addSeparatorComponents(new SeparatorBuilder().setSpacing(SeparatorSpacingSize.Large))
            .addTextDisplayComponents(
                new TextDisplayBuilder().setContent(
                    "## Civilized behavior" +
                        "\nTreat other members with respect, even if you disagree with them." +
                        "\nHate speech, harassment and discriminatory language are strictly forbidden." +
                        "\nPersonal conflicts should be resolved privately or with the help of a moderator." +
                        "\nPlease respect the decisions of moderators and administrators.",
                ),
            );

        const row = new ActionRowBuilder<ButtonBuilder>().setComponents(
            new ButtonBuilder()
                .setStyle(ButtonStyle.Link)
                .setURL(this.config.domainURL)
                .setEmoji({ id: this.emojis.logoId })
                .setLabel("Website"),
            new ButtonBuilder()
                .setStyle(ButtonStyle.Link)
                .setURL(`${this.config.domainURL}/terms`)
                .setEmoji({ id: this.emojis.logoId })
                .setLabel("ToS"),
            new ButtonBuilder()
                .setStyle(ButtonStyle.Link)
                .setURL(`${this.config.domainURL}/privacy`)
                .setEmoji({ id: this.emojis.logoId })
                .setLabel("Privacy Policy"),
        );

        return [container, row];
    }
}

File: apps/bot/src/commands/private/panels.rs
use twilight_model::channel::message::{AllowedMentions, Component, MessageFlags};

use crate::commands::prelude::*;

/// Send a panel.
pub struct PanelsCommand;

command_meta! {
    META = CommandMeta::builder("panels", "Send custom panels".into())
        .category("Private")
        .owner_only(true)
        .guilds([Config::get().support_guild_id])
        .options([
            CommandOptionBuilder::string("name", "The panel name")
                .required(true)
        ])
}

#[async_trait]
impl Command<App> for PanelsCommand {
    fn meta(&self) -> &CommandMeta {
        &META
    }

    async fn handle_command(&self, ctx: &CommandContext<'_>) -> CommandResult {
        let Some(channel) = &ctx.interaction.channel else {
            unreachable!();
        };

        ctx.defer_reply(true).await?;

        let components = match ctx.require_string_option("name")? {
            "links" => build_links_components(),
            "rules" => build_rules_components(),
            _ => {
                ctx.edit_reply(Reply::new().content("Invalid name!"))
                    .await?;
                return Ok(());
            }
        };

        ctx.http
            .create_message(channel.id)
            .flags(MessageFlags::IS_COMPONENTS_V2)
            .allowed_mentions(Some(&AllowedMentions::default()))
            .components(&components)
            .await?;

        ctx.edit_reply(Reply::new().content("Sent!")).await?;

        Ok(())
    }
}

fn build_links_components() -> Vec<Component> {
    let container = ContainerBuilder::new()
        .accent_color(Palette::PRIMARY.int)
        .add_text_display(|d| d.content(format!("## {} Orion - Links", Emojis::LOGO)))
        .add_media_gallery(|m| m.add_item_from_url(Config::BANNER_URL))
        .add_section(|s| {
            s.add_text_display(|d| {
                d.content("> Explore our hosting services or login in the dashboard.")
            })
            .set_button_accessory(|btn| {
                btn.link(Config::DOMAIN_URL)
                    .custom_emoji(Emojis::LOGO_EID, false)
                    .label("Website / Dashboard")
            })
        })
        .add_separator(|s| s.divider(true).spacing(SeparatorSpacingSize::Small))
        .add_section(|s| {
            s.add_text_display(|d| d.content("> Use the panel to manage your servers."))
                .set_button_accessory(|btn| {
                    btn.link(Config::PANEL_URL)
                        .unicode_emoji("🔗")
                        .label("Panel")
                })
        })
        .add_separator(|s| s.divider(true).spacing(SeparatorSpacingSize::Small))
        .add_section(|s| {
            s.add_text_display(|d| {
                d.content("> Read the documentation to learn how to use our hosting.")
            })
            .set_button_accessory(|btn| {
                btn.link(Config::DOCS_URL)
                    .unicode_emoji("📚")
                    .label("Documentation")
            })
        })
        .add_separator(|s| s.divider(true).spacing(SeparatorSpacingSize::Small))
        .add_section(|s| {
            s.add_text_display(|d| d.content("> Get the status of our services in real-time."))
                .set_button_accessory(|btn| {
                    btn.link(Config::STATUS_URL)
                        .custom_emoji(Emojis::PROPERTY_EID, false)
                        .label("Status")
                })
        })
        .add_separator(|s| s.divider(true).spacing(SeparatorSpacingSize::Small))
        .add_section(|s| {
            s.add_text_display(|d| d.content("> Need help? Open a ticket!"))
                .set_button_accessory(|btn| {
                    btn.link(Config::TICKETS_PANEL_URL)
                        .custom_emoji(Emojis::DISCORD_EID, false)
                        .label("Open a ticket")
                })
        });

    let legal = ContainerBuilder::new()
        .accent_color(Palette::PRIMARY.int)
        .add_text_display(|d| d.content(format!("## {} Legal Links", Emojis::LOGO)))
        .add_text_display(|d| {
            d.content(format!(
                "> Terms of Service: {}/terms\n> Privacy Policy: {}/privacy",
                Config::DOMAIN_URL,
                Config::DOMAIN_URL
            ))
        });

    vec![container.build().into(), legal.build().into()]
}

fn build_rules_components() -> Vec<Component> {
    let container = ContainerBuilder::new()
        .accent_color(Palette::PRIMARY.int)
        .add_text_display(|d| d.content(format!("## {} Orion - Rules", Emojis::LOGO)))
        .add_media_gallery(|m| m.add_item_from_url(Config::BANNER_URL))
        .add_text_display(|d| {
            d.content(
                "## Civilized behavior\n\
                Treat other members with respect, even if you disagree with them.\n\
                Hate speech, harassment and discriminatory language are strictly forbidden.\n\
                Please respect the decisions of moderators and administrators.",
            )
        })
        .add_separator(|s| s.divider(true).spacing(SeparatorSpacingSize::Large))
        .add_text_display(|d| {
            d.content(
                "## Privacy\n\
                Private messages (PMs) may not be used for advertising or spamming.\n\
                Invitations to other servers or the addition of bots are strictly forbidden.\n\
                Members' personal information must not be shared without their consent.",
            )
        })
        .add_separator(|s| s.divider(true).spacing(SeparatorSpacingSize::Large))
        .add_text_display(|d| {
            d.content(
                "## Security\n\
                Passwords and other sensitive information must not be shared.\n\
                Report any suspicious behavior or fraudulent activity to the moderators.",
            )
        })
        .add_separator(|s| s.divider(true).spacing(SeparatorSpacingSize::Large))
        .add_text_display(|d| {
            d.content(
                "## Other\n\
                Links to malicious websites or explicit content are prohibited.\n\
                Excessive use of capital letters, spam or animated emotes is prohibited.\n\
                Members may not impersonate others.",
            )
        })
        .add_separator(|s| s.divider(true).spacing(SeparatorSpacingSize::Large))
        .add_text_display(|d| {
            d.content(
                "## Civilized behavior\n\
                Treat other members with respect, even if you disagree with them.\n\
                Hate speech, harassment and discriminatory language are strictly forbidden.\n\
                Personal conflicts should be resolved privately or with the help of a moderator.\n\
                Please respect the decisions of moderators and administrators.",
            )
        });

    let row = ActionRowBuilder::new()
        .add_button(|btn| {
            btn.link(Config::DOMAIN_URL)
                .custom_emoji(Emojis::LOGO_EID, false)
                .label("Website")
        })
        .add_button(|btn| {
            btn.link(format!("{}/terms", Config::DOMAIN_URL))
                .custom_emoji(Emojis::LOGO_EID, false)
                .label("ToS")
        })
        .add_button(|btn| {
            btn.link(format!("{}/privacy", Config::DOMAIN_URL))
                .custom_emoji(Emojis::LOGO_EID, false)
                .label("Privacy Policy")
        });

    vec![container.build().into(), row.build().into()]
}
